pyk/src/pyk/kast/_ast_to_kast.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+import logging
+from functools import singledispatch
+from typing import TYPE_CHECKING
+
+from .att import EMPTY_ATT
+from .outer import KAtt, KDefinition, KFlatModule, KImport, KRequire, KSentence  # noqa: TC002
+from .outer_syntax import Att, Definition, Import, Module, Require, Sentence  # noqa: TC002
+
+if TYPE_CHECKING:
+    from typing import Any, Final
+
+    from .outer import KAst
+    from .outer_syntax import AST
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+@singledispatch
+def _ast_to_kast(ast: AST, **kwargs: Any) -> KAst:
+    _LOGGER.error(f'Unimplemented conversion for {type(ast)}')
+    return None  # type: ignore
+
+
+@_ast_to_kast.register
+def _definition_to_kdefinition(d: Definition, main_module: str) -> KDefinition:
+    modules = (_module_to_kflatmodule(m) for m in d.modules)
+    requires = (_require_to_krequire(r) for r in d.requires)
+    return KDefinition(main_module, modules, requires)
+
+
+@_ast_to_kast.register
+def _module_to_kflatmodule(m: Module) -> KFlatModule:
+    sentences = (_sentence_to_ksentence(s) for s in m.sentences)
+    imports = (_import_to_kimport(i) for i in m.imports)
+    att = _att_to_katt(m.att)
+    return KFlatModule(m.name, sentences, imports, att)
+
+
+@_ast_to_kast.register
+def _import_to_kimport(i: Import) -> KImport:
+    return KImport(i.module_name, i.public)
+
+
+@_ast_to_kast.register
+def _require_to_krequire(r: Require) -> KRequire:
+    return KRequire(r.path)
+
+
+@_ast_to_kast.register
+def _att_to_katt(att: Att) -> KAtt:
+    if not att.items:
+        return EMPTY_ATT
+    return KAtt.from_dict({'att': dict(att.items)})
+
+
+@singledispatch
+@_ast_to_kast.register
+def _sentence_to_ksentence(s: Sentence) -> KSentence:
```

Comment:
This is a placeholder for later implementation, right?

Regardless, it is not safe to trick the type checker this way. Instead consider a helper function

```python3
def _transform_sentences(sentences: Iterable[Sentence]) -> list[KSentence]:
    return []
```

which you can call in `_module_to_kflatmodule`.

Reply:
It's a placeholder in the same sense that `_ast_to_kast` is, which is that it isn't fully implemented. I can implement sentence transformations with `_sentence_to_ksentence.register`, and I think the dispatcher in `_ast_to_kast` should pass any `Sentence`s along to here.

I'm not sure what you mean by tricking the type checker. This is the same pattern as declaring `_import_to_kimport` and using it in `_module_to_kflatmodule`.

Reply:
> I'm not sure what you mean by tricking the type checker.

This:

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence:
    return None # type: ignore
```

Because here:

```python3
sentences = (_sentence_to_ksentence(s) for s in m.sentences)
```

you'll end up with a bunch of `None`s, whereas you'd expect `KSentence` instances.

So consider either weakening the signature to:

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence | None:
    return None
```

and filtering out `None`-s on the call site, or postponing the `singledispatch`-based implementation until all the cases are covered. Then, you can implement this signature as

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence:
    raise AssertionError(f'Unimplemented conversion for: {type(s)}')
```

Reply:
Ok, I see what you are saying. Well, there really is no use case for returning `None` other than as a placeholder, so I can change it to raise an `AssertionError` which wont affect anything as there aren't any tests that hit it and it isn't being used in production. It should also make it helpful as I implement the conversions and add tests for them, since the raised exception will be easier to spot than an error message from the logger.